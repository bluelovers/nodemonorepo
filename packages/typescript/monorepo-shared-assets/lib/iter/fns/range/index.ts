export type Iter = IterableIterator<number>

function range (x: number) {
  const up = (() => {
    function * from (begin = 0): Iter {
      if (begin >= x) return
      yield begin
      yield * from(begin + 1)
    }

    function to (end = 0): Iter {
      return range(end).up.from(x)
    }

    return {from, to, [Symbol.iterator]: from}
  })()

  const down = (() => {
    function * from (begin = 0): Iter {
      if (begin <= x) return
      yield begin
      yield * from(begin - 1)
    }

    function to (end = 0): Iter {
      return range(end).down.from(x)
    }

    return {from, to, [Symbol.iterator]: to}
  })()

  return {up, down, [Symbol.iterator]: up.from}
}

export default range
