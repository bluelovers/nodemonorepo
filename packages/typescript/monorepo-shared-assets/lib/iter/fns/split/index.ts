import map from '../map'

export type Iter<X> = IterableIterator<X>

export function split<X> (
  iterable: Iterable<X>,
  choose: split.func.LineChooser<X>
): Iter<split.func.Segment<X>> {
  return split.func(iterable, choose)
}

export namespace split {
  export function * func<X> (
    iterable: Iterable<X>,
    choose: func.LineChooser<X>
  ): Iter<func.Segment<X>> {
    let list = Array<X>()

    for (const item of iterable) {
      if (choose(item)) {
        yield { list, line: item, last: false }
        list = []
      } else {
        list.push(item)
      }
    }

    yield { list, line: undefined, last: true }
  }

  export namespace func {
    export type Segment<X> = Segment.NotLast<X> | Segment.Last<X>

    export namespace Segment {
      export interface Base<X> {
        readonly list: ReadonlyArray<X>
        readonly line: X | void
        readonly last: boolean
      }

      export interface NotLast<X> extends Base<X> {
        readonly line: X
        readonly last: false
      }

      export interface Last<X> extends Base<X> {
        readonly line: void
        readonly last: true
      }
    }

    export type LineChooser<X> = (x: X) => boolean
  }

  export function line<X> (
    iterable: Iterable<X>,
    line: X,
    compare: line.Comparator<X> = line.DEFAULT_COMPARATOR
  ): Iter<ReadonlyArray<X>> {
    return map(
      func(iterable, x => compare(line, x)),
      x => x.list
    )
  }

  export namespace line {
    export type Comparator<X> = (a: X, b: X) => boolean
    export const DEFAULT_COMPARATOR = Object.is
  }
}

export default split
