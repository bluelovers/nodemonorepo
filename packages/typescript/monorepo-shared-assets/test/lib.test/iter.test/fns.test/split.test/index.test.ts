import {iter} from '../../../../../index'
import range = iter.fns.range
import split = iter.fns.split

describe('split by function', () => {
  const choose = (x: number) => x % 5 === 0
  const getArray = (x: number) => Array.from(split.func(range(1).up.to(x), choose))
  const getDefaultArray = (x: number) => Array.from(split(range(1).up.to(x), choose))

  it('when tail is not empty', () => {
    expect(getArray(18)).toMatchSnapshot()
    expect(getArray(18)).toEqual(getDefaultArray(18)) // default
  })

  it('when tail is empty', () => {
    expect(getArray(16)).toMatchSnapshot()
    expect(getArray(16)).toEqual(getDefaultArray(16)) // default
  })
})

describe('split by line', () => {
  const iterable = [0, null, 1, 2, null, 3, 4, 5, null, 6, 7, 8, 9]
  const line = null

  const getArray = (compare?: split.line.Comparator<any>) =>
    Array.from(split.line(iterable, line, compare))

  it('using default comparator', () => {
    expect(getArray()).toMatchSnapshot()
  })

  it('using custom comparator', () => {
    expect(getArray((a, b) => a === b)).toMatchSnapshot()
  })
})
