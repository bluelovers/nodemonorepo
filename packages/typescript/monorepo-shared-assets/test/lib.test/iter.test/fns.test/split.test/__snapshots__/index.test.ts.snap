// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`split by function when tail is empty 1`] = `
Array [
  Object {
    "last": false,
    "line": 5,
    "list": Array [
      1,
      2,
      3,
      4,
    ],
  },
  Object {
    "last": false,
    "line": 10,
    "list": Array [
      6,
      7,
      8,
      9,
    ],
  },
  Object {
    "last": false,
    "line": 15,
    "list": Array [
      11,
      12,
      13,
      14,
    ],
  },
  Object {
    "last": true,
    "line": undefined,
    "list": Array [],
  },
]
`;

exports[`split by function when tail is not empty 1`] = `
Array [
  Object {
    "last": false,
    "line": 5,
    "list": Array [
      1,
      2,
      3,
      4,
    ],
  },
  Object {
    "last": false,
    "line": 10,
    "list": Array [
      6,
      7,
      8,
      9,
    ],
  },
  Object {
    "last": false,
    "line": 15,
    "list": Array [
      11,
      12,
      13,
      14,
    ],
  },
  Object {
    "last": true,
    "line": undefined,
    "list": Array [
      16,
      17,
    ],
  },
]
`;

exports[`split by line using custom comparator 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    1,
    2,
  ],
  Array [
    3,
    4,
    5,
  ],
  Array [
    6,
    7,
    8,
    9,
  ],
]
`;

exports[`split by line using default comparator 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    1,
    2,
  ],
  Array [
    3,
    4,
    5,
  ],
  Array [
    6,
    7,
    8,
    9,
  ],
]
`;
